helix-term/src/ui/explorer.rs

```diff
@@ -0,0 +1,1527 @@
+            path: path.join(entry.file_name()),
+        }
+    })
+}
+
+#[derive(Clone, Debug)]
+enum PromptAction {
+    CreateFileOrFolder,
+    RemoveFolder,
+    RemoveFile,
+    RenameFile,
+}
+
+#[derive(Clone, Debug, Default)]
+struct State {
+    focus: bool,
+    open: bool,
+    current_root: PathBuf,
+    area_width: u16,
+}
+
+impl State {
+    fn new(focus: bool, current_root: PathBuf) -> Self {
+        Self {
+            focus,
+            current_root,
+            open: true,
+            area_width: 0,
+        }
+    }
+}
+
+struct ExplorerHistory {
+    tree: TreeView<FileInfo>,
+    current_root: PathBuf,
+}
+
+pub struct Explorer {
+    tree: TreeView<FileInfo>,
+    history: Vec<ExplorerHistory>,
+    show_help: bool,
+    show_preview: bool,
+    state: State,
+    prompt: Option<(PromptAction, Prompt)>,
+    #[allow(clippy::type_complexity)]
+    on_next_key: Option<Box<dyn FnMut(&mut Context, &mut Self, &KeyEvent) -> EventResult>>,
+    column_width: u16,
+}
+
+impl Explorer {
+    pub fn new(cx: &mut Context) -> Result<Self> {
+        let current_root = std::env::current_dir()
+            .unwrap_or_else(|_| "./".into())
+            .canonicalize()?;
+        Ok(Self {
+            tree: Self::new_tree_view(current_root.clone())?,
+            history: vec![],
+            show_help: false,
+            show_preview: false,
+            state: State::new(true, current_root),
+            prompt: None,
+            on_next_key: None,
+            column_width: cx.editor.config().explorer.column_width as u16,
+        })
+    }
+
+    #[cfg(test)]
+    fn from_path(root: PathBuf, column_width: u16) -> Result<Self> {
+        Ok(Self {
+            tree: Self::new_tree_view(root.clone())?,
+            history: vec![],
+            show_help: false,
+            show_preview: false,
+            state: State::new(true, root),
+            prompt: None,
+            on_next_key: None,
+            column_width,
+        })
+    }
+
+    fn new_tree_view(root: PathBuf) -> Result<TreeView<FileInfo>> {
+        let root = FileInfo::root(root);
+        Ok(TreeView::build_tree(root)?.with_enter_fn(Self::toggle_current))
+    }
+
+    fn push_history(&mut self, tree_view: TreeView<FileInfo>, current_root: PathBuf) {
+        self.history.push(ExplorerHistory {
+            tree: tree_view,
+            current_root,
+        });
+        const MAX_HISTORY_SIZE: usize = 20;
+        Vec::truncate(&mut self.history, MAX_HISTORY_SIZE)
+    }
+
+    fn change_root(&mut self, root: PathBuf) -> Result<()> {
+        if self.state.current_root.eq(&root) {
+            return Ok(());
+        }
+        let tree = Self::new_tree_view(root.clone())?;
+        let old_tree = std::mem::replace(&mut self.tree, tree);
+        self.push_history(old_tree, self.state.current_root.clone());
+        self.state.current_root = root;
+        Ok(())
+    }
+
+    pub fn reveal_file(&mut self, path: PathBuf) -> Result<()> {
```

Comment:
Is there any reason why this is called `reveal_file` rather than `open_file`?

Reply:
Because `reveal_file` means to reveal the current file in the file explorer, not opening the file as the file has been opened already.